### ISSUE TITLE
Goal Files

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#4992]
### Purpose

The purpose of this ticket is to discuss how to bring over the watch mode workflow presently available at Jenga to dune.

### Jenga

I'll first describe what I know about Jenga's system. Feel free to correct me here.

Jenga may be started in watch mode with a "start file". This start file is essentially a text file containing a list of targets that Jenga should be build. This file is edited both by users and tools.

Some problems I see with Jenga's approach:

- One must remember to create the start file before running Jenga. It's usually fine to restart the build, but it is a problem when one expects the tooling to edit this file automatically.
- It's not possible to configure any other aspect at runtime. E.g. one cannot set an environment variable.
- For automated use, one does not get notified when a part of the build completes. This is an acceptable limitation, as it's only useful for some UI features.

### Dune

For dune, we want to solve the issues above and in addition, we have one more constraint: we prefer not to add yet another type of file. What we can do is add a goals stanza to the workspace file. We'll also need some RPC calls to edit this file.

cc @jeremiedimino @aalekseyev @cwong-ocaml @tmattio

**Comment by jeremiedimino** [comment]
> One must remember to create the start file before running Jenga. It's usually fine to restart the build, but it is a problem when one expects the tooling to edit this file automatically.

At Jane Street, the editor does this. More precisely, we give a few buttons to the user (which are emacs key bindings for us, but let's call them buttons):
  - build the default alias in the current directory
  - build runtest in the current directory
  - and a few other things
  - build the current feature (= current branch), where the system tries to guess the set of targets to build based on the feature diff
  
The first time the user presses one of these buttons, the editor:
- creates a file `jenga/start/jbuild`
- calls `dune build @jenga/start/build`

The second time the user presses one of these buttons, the editor extend the existing `jenga/start/jbuild` file with the relevant targets.

For instance, if the users press the button to build the default alias in `lib/base/`, the editor produces the following `jbuild` file:

```
(alias
 ((name build)
  (deps ((alias %{root}/lib/base/DEFAULT)))))
```

> It's not possible to configure any other aspect at runtime. E.g. one cannot set an environment variable.

That's something we'd like as well, but never took the time to implement.

**Comment by rgrinberg** [comment]
Thanks for clarifying Jeremie.

> The second time the user presses one of these buttons, the editor extend the existing jenga/start/jbuild file with the relevant targets.

I think this is the main part that we can't copy. we aren't yet in a position to invoke dune for the user. It's something that we'll get to, but for now we should just work around it.  In any case, we already have watching infrastructure, so it should be trivial for dune to know when this file is created.

**Comment by rgrinberg** [comment]
This was discussed by the team. A quick summary of the discussion:

* We prefer to use the sexp based syntax for the goal files rather than the command line syntax.
* It was brought up that the goal file should  be interpreted by dune, but not necessarily checked in anywhere.
* It was suggested to move the goal file to `_build`. This suggestion was rejected because we don't want the user writing anything into that directory.

If we adopt jenga's scheme of having no convention for this file, we require the user to remember a pretty long command line invocation for the most common use case in watch mode. Ideally, `dune build -w` should just "do the right thing".

Unfortunately, I don't see an alternative to this other than introducing some sort of convention in dune itself.

One alternative would be to go back to the _build idea, but wrap it with porcelain that makes the interaction with _build more invisible. For example, introduce a `$ dune goal file` sub command that will print the path to the goal file likely in _build. Or add `$ dune goal edit` that will launch the editor with the goal file ready.

@emillon @snowleopard it's back to the drawing board unfortunately.

**Comment by snowleopard** [comment]
> we require the user to remember a pretty long command line invocation for the most common use case in watch mode. Ideally, `dune build -w` should just "do the right thing".

@rgrinberg I don't understand this bit: what's so long or otherwise bad about `dune build @goal -w` where `@goal` is this alias? This works today and requires no special support in Dune.

**Comment by snowleopard** [comment]
Actually, this doesn't even require recursive aliases:

* The `dune` file at the root defines alias `@goal` that depends on `(alias my-special-ignored-dir/goal)`.
* The `my-special-ignored-dir/dune` file defines the actual `@goal` alias, which is easily editable because it isn't committed.

The user can now simply `dune build @goal -w` and it all seems to work.

**Comment by rgrinberg** [comment]
> The my-special-ignored-dir/dune file defines the actual @goal alias, which is easily editable because it isn't committed.

This is what creates the problem for creating the good editor tooling like you have at JST though. There's no standard `my-special-ignored-dir` that we can bless so that the editor can create/edit automatically.

**Comment by snowleopard** [comment]
So, then let's be clearer about the problem we are solving. In my mind, it isn't a technical problem. Dune has enough features already, so you can hook up any editor to it to achieve the desired "goal file" behaviour. It's a social/standards problem and so it doesn't necessarily require any changes in Dune. It can merely be documented as a convention. What's wrong with that?

We can of course encode this convention in Dune, somehow, perhaps by providing some fancy shortcuts that help you save a few characters, but that comes with some complexity and maintenance costs.

To be clear, I'm not strongly arguing against encoding this convention in Dune: the key point I'm making is that it's unnecessary.

**Comment by rgrinberg** [comment]
Yes, you're right that the problem isn't technical. The goal is to encourage this workflow as much as possible. In my experience, users do not read documentation and the adoption of any non-essential features essentially does not happen until the feature works in a seamless way. So what we're looking for is ways to simplify all of this and just make the most common use case of `dune build -w` + vscode "just work" with goal files.

**Comment by snowleopard** [comment]
Good that we're on the same page!

One thing I don't understand is: why is CLI important for this feature? Shouldn't VSCode, Emacs etc. invoke Dune themselves with the right parameters?

**Comment by cwong-ocaml** [comment]
>  In particular, we don't have a good story when the user tries to run `dune` manually while the `dune` executed by the editor is running. I suppose we could work on improving that first.

I think the most seamless thing to do would be to designate the already-running dune process as a "server" and then have the new dune just print out information from that. In Jenga, we have `jenga monitor`, which mirrors the progress display; pretty much all the machinery is already in place to be able to mirror progress and diagnostics.

**Comment by snowleopard** [comment]
> > One thing I don't understand is: why is CLI important for this feature? Shouldn't VSCode, Emacs etc. invoke Dune themselves with the right parameters?
> 
> That's something we'd like to do but we aren't ready for yet. In particular, we don't have a good story when the user tries to run `dune` manually while the `dune` executed by the editor is running. I suppose we could work on improving that first.

That seems like something we'll need to solve anyway. Having concurrent instances of Dune messing around with each other's files isn't great. I guess we also need a way to stop a currently running Dune for a given workspace, e.g. via `dune stop`.

**Comment by Alizter** [comment]
This is something that I want to be able to have for TUI.

**Comment by rgrinberg** [comment]
I think we need to understand how it's going to understand without any tui first

**Comment by Alizter** [comment]
Yes, I should clear up that this should be a general feature, but is something I can make easier to interact with using TUI.

**Comment by rgrinberg** [comment]
I think we'll just copy jenga here.
